### DISCUSSION TITLE
swc_ecma_lints overhead

### DISCUSSION THREAD

**Discussion by sapphi-red** [rollup/rollup#5205] [Ideas]
I found that replacing this part with `Ok(progam)` trims 300-400ms from the current master branch (https://github.com/rollup/rollup/commit/ee0639e711672b552fcd2928a700bc5541a564f0) on the benchmark I used in https://github.com/rollup/rollup/pull/5201.
https://github.com/rollup/rollup/blob/ee0639e711672b552fcd2928a700bc5541a564f0/rust/parse_ast/src/error_emit.rs#L71-L101

I guess this is because acorn ran these lints while running the parser (so it's a 1 pass), while SWC needs to run these separately (2 pass).

The current parse algorithm is a 3 pass.
1. parse with SWC
2. lint with SWC
3. AST binary transformation

If we made the lint work while doing the binary transformation, we can make it as 2 pass and will reduce the time.
Or given that the lint is no longer part of the parse, maybe it can be simply removed?

**Comment by lukastaegert** [comment]
That is a very good point that I completely forgot about. I think we need these validations at some point in the Rollup pipeline as subsequent logic like tree-shaking may rely on them. But it is only four rules:
* const_assign,
* duplicate_bindings,
* duplicate_exports,
* no_dupe_args

While we could add them to the binary transform with reasonable effort (and probably will eventually), I would want to avoid this until we moved scope analysis to Rust. Instead, I think we can easily handle them in `initialise()` handlers on the JavaScript AST nodes. That would mean we no longer throw in `this.parse`, but maybe this is acceptable for now. The performance gains very much seem to be worth it.
